domain/user/state/state.go

```diff
@@ -0,0 +1,395 @@
+	if err != nil {
+		return fmt.Errorf("cannot validate user uuid %q: %w", uuid, err)
+	}
+	err = creatorUuid.Validate()
+	if err != nil {
+		return fmt.Errorf("cannot validate creator uuid %q: %w", creatorUuid, err)
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		err = addUser(ctx, tx, uuid, user, creatorUuid)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return setActivationKey(ctx, tx, uuid, activationKey)
+	})
+	if err != nil {
+		return fmt.Errorf("cannot add user %q with activation key: %w", user.Name, err)
+	}
+
+	return nil
+}
+
+// GetUser will retrieve the user specified by name from the database where
+// the user is active and has not been removed. If the user does not exist
+// or is deleted an error that satisfies usererrors.NotFound will be
+// returned.
+func (st *State) GetUser(ctx context.Context, uuid user.UUID) (user.User, error) {
+	err := uuid.Validate()
+	if err != nil {
+		return user.User{}, fmt.Errorf("cannot validate user uuid %q: %w", uuid, err)
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return user.User{}, fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	var domainUser user.User
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		domainUser, err = getUser(ctx, tx, uuid)
+		return errors.Trace(err)
+	})
+
+	return domainUser, errors.Trace(err)
+}
+
+// RemoveUser marks the user as removed. This obviates the ability of a user
+// to function, but keeps the user retaining provenance, i.e. auditing.
+// RemoveUser will also remove any credentials and activation codes for the
+// user. If no user exists for the given name then an error that satisfies
+// usererrors.NotFound will be returned.
+func (st *State) RemoveUser(ctx context.Context, uuid user.UUID) error {
+	err := uuid.Validate()
+	if err != nil {
+		return fmt.Errorf("cannot validate user uuid %q: %w", uuid, err)
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	removeUserQuery := "UPDATE user SET removed = true WHERE uuid = ?"
+
+	return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		result, err := tx.ExecContext(ctx, removeUserQuery, uuid.String())
+		if num, err := result.RowsAffected(); err != nil {
+			return usererrors.NotFound
+		} else if num != 1 {
+			return fmt.Errorf("expected to remove 1 user, but removed %d", num)
+		}
+		return errors.Trace(err)
+	})
+}
+
+// SetActivationKey removes any active passwords for the user and sets the
+// activation key. If no user is found for the supplied name an error
+// is returned that satisfies usererrors.NotFound.
+func (st *State) SetActivationKey(ctx context.Context, uuid user.UUID, activationKey string) error {
+	err := uuid.Validate()
+	if err != nil {
+		return fmt.Errorf("cannot validate user uuid %q: %w", uuid, err)
+	}
+
+	removed, err := st.isRemoved(ctx, uuid)
```

Comment:
These removed checks need to happen in the same transaction as removing the password has and set activation key.

My advice would be to:
- Remove these checks and add the checks to both removePasswordHash() and setActivationKey()
- If we are going to error on removed user now because of the UUID we should create a new strongly typed error to return and update all of our contracts so we can give proper feedback to the caller.